OpenRA.Mods.Common/Traits/BotModules/SendUnitToAttackBotModule.cs

```diff
@@ -0,0 +1,186 @@
+
+		public override object Create(ActorInitializer init) { return new SendUnitToAttackBotModule(init.Self, this); }
+	}
+
+	public class SendUnitToAttackBotModule : ConditionalTrait<SendUnitToAttackBotModuleInfo>, IBotTick
+	{
+		readonly World world;
+		readonly Player player;
+		readonly Predicate<Actor> unitCannotBeOrdered;
+		readonly Predicate<Actor> unitCannotBeOrderedOrIsBusy;
+		readonly Predicate<Actor> isInvalidActor;
+		int minAssignRoleDelayTicks;
+		Player targetPlayer;
+		int desireIncreased;
+
+		public SendUnitToAttackBotModule(Actor self, SendUnitToAttackBotModuleInfo info)
+		: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+			isInvalidActor = a => a == null || a.IsDead || !a.IsInWorld || a.Owner != targetPlayer;
+			unitCannotBeOrdered = a => a == null || a.IsDead || !a.IsInWorld || a.Owner != player;
+			unitCannotBeOrderedOrIsBusy = a => unitCannotBeOrdered(a) || (!a.IsIdle && a.CurrentActivity is not FlyIdle);
+			desireIncreased = 0;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			// Avoid all AIs reevaluating assignments on the same tick, randomize their initial evaluation delay.
+			minAssignRoleDelayTicks = world.LocalRandom.Next(0, Info.ScanTick);
+		}
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			if (--minAssignRoleDelayTicks <= 0)
+			{
+				minAssignRoleDelayTicks = Info.ScanTick;
+
+				var attackdesire = 0;
+				var actors = world.ActorsWithTrait<IPositionable>().Select(at => at.Actor).Where(a =>
+				{
+					if (Info.ActorTypesAndAttackDesire.ContainsKey(a.Info.Name) && !unitCannotBeOrderedOrIsBusy(a))
+					{
+						attackdesire += Info.ActorTypesAndAttackDesire[a.Info.Name];
+						return true;
+					}
+
+					return false;
+				}).ToList();
+
+				if (actors.Count == 0)
+					desireIncreased = 0;
+				else
+					desireIncreased += Info.AttackDesireIncreasedPerScan;
+
+				if (desireIncreased + attackdesire < 100)
+					return;
+
+				// Randomly choose target player to attack
+				var targetPlayers = world.Players.Where(p => Info.ValidRelationships.HasRelationship(p.RelationshipWith(player)) && p.WinState != WinState.Lost).ToList();
+				if (targetPlayers.Count == 0)
+					return;
+
+				targetPlayer = targetPlayers.Random(world.LocalRandom);
+
+				var targets = world.Actors.Where(a =>
+				{
+					if (isInvalidActor(a))
+						return false;
+
+					var t = a.GetAllTargetTypes();
+
+					if (!Info.ValidTargets.Overlaps(t) || Info.InvalidTargets.Overlaps(t))
+						return false;
+
+					var hasModifier = false;
+					var visModifiers = a.TraitsImplementing<IVisibilityModifier>();
+					foreach (var v in visModifiers)
+					{
+						if (v.IsVisible(a, player))
+							return true;
+
+						hasModifier = true;
+					}
+
+					return !hasModifier;
+				});
+
+				switch (Info.AttackDistance)
+				{
+					case AttackDistance.Closest:
+						targets = targets.OrderBy(a => (a.CenterPosition - actors[0].CenterPosition).HorizontalLengthSquared);
```

Comment:
there should be some randomness here. This will fail spectacularly on larger maps with more players

Reply:
Ummm, don't get it. You can use `AttackDistance.Random` for randomness.

Reply:
so everything will need to use `AttackDistance.Random`, the `Farthest` option is unusable. There's a lot of scenarios where it will glitch out, and even more where it will act in very illogical manner. It should be more like a random with weight, as in there's a higher change to target units farther in the back, but not at the very back. We also need to keep in mind that this isn't using pathfinder distance.

The Closest is serviceable, but won't be nearly good enough. Many units needs some sort of auto-targeting for switching to the target that actually is next to them

Reply:
> the `Farthest` option is unusable

I have 2 usecases and they works pretty well so far.

1. Demo truck has a large damage range even when it is destroyed, so get as deep into enemy base as possible is a good idea. Even if player knows how the bot module works, it is impossible for player to stop all autotarget from units and defense when Demo truck get into bases.

2. Subtank can get through enemy defense line without harm, so attack furthest enemy building will be a effect way of force player invest more on defend the backlines.

> We also need to keep in mind that this isn't using pathfinder distance.

well if performance allow, we can use pathfind distance.

> The Closest is serviceable, but won't be nearly good enough. Many units needs some sort of auto-targeting for switching to the target that actually is next to them

Then I think it out of the scope of this simple module, and we can do autotarget on unit for AI after auto target is refactored.

Reply:
>  in there's a higher change to target units farther in the back, but not at the very back. 

Actually even this is not enough at all if we put it to detailed consideration, we also need to scan dangerous actors in radius, planning the attack route and abort mission/fight on halfway ambush, obviously it should reuse some of the functions in SquadManager.

The purpose of this module is to use a simple/light-weight way for those units before the https://github.com/OpenRA/OpenRA/issues/9021 is finally implemented.